### ISSUE TITLE
Using the transport

### ISSUE THREAD

**Issue by 4lbert** [rustyhorde/twist#13]
Hi,

Ever since finding this library I've wanted to use it in a websocket server that does not follow a request/response structure (my first question on the topic: https://github.com/rustyhorde/twist/issues/8).

At that time I did not understand tokio all that much and I only got it to read messages from clients, but I never managed to send websockets to the connected client. Since I still don't understand tokio that well I'm wondering if anyone could help me with this. It also seems reasonable for an example of this to be provided in the documentation given that this seems like a common use case given the nature of websockets.

I feel as if I'm close to figuring it out, but I just can't take it all the way. This is how far I came:

```rust
extern crate futures;
extern crate tokio_core;
extern crate tokio_proto;
extern crate twist;

use std::io::{Read, Write};
use tokio_core::io::{Io, copy};
use tokio_core::reactor::Core;
use tokio_core::net::{TcpStream, TcpListener};
use tokio_proto::pipeline::ServerProto;
use futures::{Future, Stream, Sink};
use twist::{WebSocketProtocol, TwistCodec, WebSocketFrame};

fn main() {
    let mut core = Core::new().unwrap();
    let handle = core.handle();

    let address = "127.0.0.1:3012".parse().unwrap();
    let listener = TcpListener::bind(&address, &handle).unwrap();
    let server = listener.incoming().for_each(move |(socket, addr)| {
        let (mut ws_sink, ws_stream) = WebSocketProtocol::default().bind_transport(socket)
            .unwrap().split();

        let process_connection = ws_stream.for_each(move |frame| {
            println!("{} >>> \"{}\"",
                addr,
                std::str::from_utf8(frame.base().unwrap().application_data().unwrap()).unwrap()
            );
            Ok(())
        });
        handle.spawn(process_connection.map_err(|_| ()));

        // let process_send = ws_sink.send(WebSocketFrame::default());
        // handle.spawn(process_send);

        // This is how I would think it would be done, but this gives errors, more specifically
        // it says that the spawn function expected a futures::stream::SplitSink but found a ()
        // and a std::io::Error but found a ().

        Ok(())
    });

    core.run(server).unwrap();
}
```

Basically I figured out that I need to create a TcpListener and then bind the transport to the socket with `WebSocketProtocol::default().bind_transport(socket).unwrap()` but I can't figure out how to send websocket-frames through the transport (seems easy, but I just can't figure it out).

Really hope that anyone could help me figuring this out, would really appreciate it!

**Comment by CraZySacX** [comment]
You can't use the transport from tokio-proto if you want a non-request/response based server.   Even the streaming protocol is intended for processing the streamed bodies in a request/response architecture.

If you want to continue down the path you are on, your `process-send`  should be wrapped in something like `future::ok` so that the future completes.

To implement a fully bidirectional websocket server I would recommend only using the codec from twist, tracking the state of each incoming connection manually (i.e. has handshake occurred, etc.), and possibly surrounding the send with mpsc so you can send from the server whenever you need.

**Comment by 4lbert** [comment]
Thank you for the explanation! Implementing a bidirectional websocket server using only the codec and a mpsc sounds like the right way to go. However, it sounds hard (for me), but I will give it a try!
